streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java

```diff
@@ -217,16 +218,32 @@ public <K, V> void send(final String topic,
                     valueClass),
                 exception);
         } catch (final Exception exception) {
+            final ProductionExceptionHandler.SerializationExceptionOrigin origin;
+            if (keyBytes == null) {
+                origin = ProductionExceptionHandler.SerializationExceptionOrigin.KEY;
+            } else {
+                origin = ProductionExceptionHandler.SerializationExceptionOrigin.VALUE;
+            }
```

Comment:
I think, it would be better to have something like:
```java
try {
    keyBytes = keySerializer.serialize(topic, headers, key);
} catch (final ClassCastException exception) {
    handleClassCastException(...)
} catch (final Exception exception) {
   handleException(ProductionExceptionHandler.SerializationExceptionOrigin.KEY, ...)
}

try {
    valBytes = valueSerializer.serialize(topic, headers, value);
} catch (final ClassCastException exception) {
    handleClassCastException(...)
} catch (final Exception exception) {
   handleException(ProductionExceptionHandler.SerializationExceptionOrigin.VALUE, ...)
}
```

Reply:
Done.

I had to remove the `final` from `keyBytes` and `valueBytes` and initialize them when moving the `ClassCastException` code to a dedicated method

Reply:
If you want to keep the `final` and make the code a bit better readable, you could rename `handleClassCastException()` to `createStreamsExceptionForClassCastException()` and return the `StreamsException` from there. Then you can do
```java
} catch (final ClassCastException exception) {
            throw createStreamsExceptionForClassCastException()(
                topic,
                key,
                value,
                keySerializer,
                valueSerializer,
                exception);
```
With this it is clear that the `catch`-branch terminates the method.

Reply:
The update has been pushed!